pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -103,11 +104,23 @@ public BrokerResponseNative reduceOnDataTable(BrokerRequest brokerRequest,
       return brokerResponseNative;
     }
 
-    QueryContext queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
-    DataTableReducer dataTableReducer = ResultReducerFactory.getResultReducer(queryContext);
+    QueryContext serverQueryContext = BrokerRequestToQueryContextConverter.convert(serverBrokerRequest);
+    DataTableReducer dataTableReducer = ResultReducerFactory.getResultReducer(serverQueryContext);
     dataTableReducer.reduceAndSetResults(rawTableName, cachedDataSchema, dataTableMap, brokerResponseNative,
         new DataTableReducerContext(_reduceExecutorService, _maxReduceThreadsPerQuery, reduceTimeOutMs,
             _groupByTrimThreshold), brokerMetrics);
+    QueryContext queryContext;
+    if (brokerRequest == serverBrokerRequest) {
+      queryContext = serverQueryContext;
+    } else {
+      queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
+    }
+
+    GapfillUtils.GapfillType gapfillType = GapfillUtils.getGapfillType(queryContext);
+    if (gapfillType != null) {
+      GapfillProcessor gapfillProcessor = new GapfillProcessor(queryContext, gapfillType);
+      gapfillProcessor.process(brokerResponseNative);
+    }
```

Comment:
No need to check gapfill type when server request is the same as broker request
```suggestion
    if (brokerRequest == serverBrokerRequest) {
      queryContext = serverQueryContext;
    } else {
      queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
      GapfillUtils.GapfillType gapfillType = GapfillUtils.getGapfillType(queryContext);
      if (gapfillType != null) {
        GapfillProcessor gapfillProcessor = new GapfillProcessor(queryContext, gapfillType);
        gapfillProcessor.process(brokerResponseNative);
      }
    }
```

Reply:
Fixed